src/topology/homotopy/pi_n.lean

```diff
@@ -0,0 +1,364 @@
+
+    /--
+    The constant homotopy at f
+    -/
+    @[refl] def refl (f : pre_π n x) : f ~ f :=
+      continuous_map.homotopy_rel.refl _ _
+
+    /--
+    Reversing of a homotopy
+    -/
+    @[symm] def symm : f ~ g -> g ~ f :=
+      continuous_map.homotopy_rel.symm
+
+    /--
+    Concatenation of homotopies by using them along `[0,2⁻¹]` and `[2⁻¹,1] `
+    -/
+    @[trans] def trans : f ~ g -> g ~ h -> f ~ h :=
+      continuous_map.homotopy_rel.trans
+
+    /--
+    Eval an homotopy at an intermediate point, yielding a `pre_π`
+    -/
+    def eval (t:I) : pre_π n x :=
+      { to_fun := H.to_homotopy.curry t,
+        continuous_to_fun := by {continuity},
+        boundary := λ y H, (begin
+          simp, rw continuous_map.homotopy_rel.eq_fst,
+          apply f.boundary, assumption, assumption,
+        end), }
+
+    @[simp] lemma eval0 : H.eval 0 = f :=
+      begin ext, unfold eval,
+        simp only [mk_apply, continuous_map.homotopy_with.apply_zero,
+          continuous_map.coe_mk, continuous_map.homotopy.curry_apply,
+          continuous_map.homotopy_with.coe_to_homotopy],
+        reflexivity,
+      end
+
+    @[simp] lemma eval1 : H.eval 1 = g :=
+      begin ext, unfold eval,
+        simp only [mk_apply, continuous_map.coe_mk,
+          continuous_map.homotopy_with.apply_one, continuous_map.homotopy.curry_apply,
+          continuous_map.homotopy_with.coe_to_homotopy],
+        reflexivity,
+      end
+
+    @[simp] lemma at0 (y : I^n) : H (fin.cons 0 y) = f y :=
+      begin convert H.apply_zero y, simp end
+
+    @[simp] lemma at1 (y : I^n) : H (fin.cons 1 y) = g y :=
+      begin convert H.apply_one y, simp end
+
+    @[simp] lemma comp0 : H ∘ fin.cons 0 = f :=
+      funext (at0 H)
+
+    @[simp] lemma comp1 : H ∘ fin.cons 1 = g :=
+      funext (at1 H)
+
+    end
+
+    instance inhabited : inhabited
+      ( (const:pre_π n x).homotopy const ) :=
+      { default := { to_fun := λ_,x,
+        continuous_to_fun := by continuity,
+        map_zero_left' := λ_,rfl,
+        map_one_left' := λ_,rfl,
+        prop' := λ_ _ _,⟨rfl,rfl⟩ } }
+  end homotopy
+
+  /--
+  Homotopic relation between `pre_π`s
+  -/
+  def homotopic (f0 f1 : pre_π n x) : Prop := nonempty (f0 ~ f1)
+  namespace homotopic
+    @[refl] lemma refl (f : pre_π n x) : homotopic f f :=
+      ⟨homotopy.refl f⟩
+
+    @[symm] lemma symm {f g : pre_π n x} (h : homotopic f g) : homotopic g f :=
+      h.map homotopy.symm
+
+    @[trans] lemma trans {f g h : pre_π n x} (h0 : homotopic f g) (h1 : homotopic g h)
+      : homotopic f h :=
+      h0.map2 homotopy.trans h1
+    instance setoid (n : nat) (x : X) : setoid (pre_π n x) :=
+      ⟨homotopic, ⟨homotopic.refl, λ _ _, homotopic.symm, λ _ _ _, homotopic.trans⟩⟩
+  end homotopic
+
+end pre_π
+
+/--
+The nth homotopy group at x defined as the quotient of base preserving functions up to homotopic
+equivalence
+-/
+def π (n : nat) (x : X) := quotient (pre_π.homotopic.setoid n x)
```

Comment:
This is way too generic of a name to be in the root namespace. Would `homotopy_group` be any better?

Consider renaming `pre_π` as well.